docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+# ADR 011: kURL Add-on Operator
+
+## Context
+
+kURL today is a set of bash scripts that configure and install a Kubernetes cluster in addition to a set of add-on components. The components are installed in sequence by applying highly customized Kubernetes YAML resources. The current architecture makes kURL hard to develop, maintain, and support.
+
+As an amalgam of bash scripts, the code base suffers from low cohesion and high complexity which results in the code being difficult to read and reason, and difficult to test. The installation is applied sequentially and each component is highly dependent on each other with no scopes to limit access and enforce separation. The project is architecturally a monolith without any functional separation of components and concerns.
+
+The installation of all components is done procedurally in the script. When there is a failure, it is difficult to understand what components have failed and how that affects the installation process as a whole.
+
+The components are installed in a proprietary fashion by applying highly customized Kubernetes YAML resources. They are hard to customize as the spec is proprietary and hard to extend because each customization must be passed through configuration to the end-user. Changes to this spec must be documented and are less discoverable. This method is not supported by the upstream which makes it difficult to pin addon installation issues on the upstream project and difficult to request support from the upstream.
+
+
+## Decision
+
+We will develop a Kubernetes [Operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) in _Go_ that will install, configure and manage add-on applications in the cluster. The operator will extend the Kubernetes API through Custom Resource Definitions (CRD) and will allow users to declaritively manage add-ons with a Custom Resource (CR). 
+
+## Solution
+
+The operator will be implemented with the following objectives in mind:
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
+- The installation mechanism for add-ons will be through the use of Helm Charts.
+- By default the add-on configuration will be opinionated by us but we will enable the end-user to override the configuration for maximum flexibility. Vendors and users will be able to override the configuration using a standard [Helm values](https://helm.sh/docs/chart_template_guide/values_files/) file.
```

Comment:
I think we need either call down the complexities that we will have within just for we are all aware of:
- a) Users might add invalid data we will need to validate those inputs
- b) Upper version of HelmCharts might need new values or no longer work with values informed previously which means that is a scenario that we need to handle into upgrades
- c) Some AddOns configuration might not work well with other AddOns configuration, that is something that is very complex to prevent or test out. We either will not able to test all possible values combinations.

Reply:
I agree. Let me add one thing: we are doing a lot of effort to ensure we can guarantee seamless cluster upgrades and IMHO to expose `values.yaml` would render the same situation we have nowadays. We won't be able to test all possible combinations during upgrades.

Reply:
I can foresee the pitfalls with allowing users to pass/override the values for an add-on. However, there is a need to allow end-users to override _certain_ add-on configs for e.g. Rook (filters for OSD devices) and containerd (config for GPU support) to name a few. We can either i) restrict what configs can be overidden or ii) allow unmitigated access to override `values` but with the caveat that users "proceed at their own risk".

Reply:
@rrpolanco, both @ricardomaraschini and I seems agree with your perspective `We can either i) restrict what configs can be overidden or ii) allow unmitigated access to override values but with the caveat that users "proceed at their own risk".`. 

Considering the inherent risks and the specific business requirements that kURL aims to address - such as facilitating seamless upgrades, automating workarounds, and minimizing complexities for users - it's crucial to underscore the potential challenges that arise when users are permitted to add any chart value.

This flexibility comes with its own set of complexities: verifying the validity of these values, ensuring compatibility between various solutions, and handling possible conflicts. For instance, accommodating both OpenEBS and Rook in the same cluster with any given configuration could pose challenges and that we cannot test all possible combinations. Furthermore, we need to address what happens during upgrades when a particular value is no longer valid.

Balancing user freedom and system stability is a delicate act, and it's paramount that we consider these factors to avoid complicating the user experience or introducing instability to the system.

Then, I think we can propose this solution and call out all complexities and problems inherent to this approach as add an alternative option where we limit the possible options within the pros and cons.

Reply:
kURL cannot gatekeep add-on configuration or else we will fall into the same scenario with the current kURL offering which is that we have to release a new feature every time a vendor/end-user wants to configure something outside of the normal use case. E.g. configuring GPU support for containerd.
I think a reasonable approach is to document our opinionated (and supported) configuration (`values.yaml`) for each add-on and allow vendors/users to overwrite them with the warning that their applied config may not be supported (i.e. they can proceed but at their own risk).

Reply:
I am totally OK with that.
I just think that in the ADR we should explicitly describe the limitations and potential problems. Just that. we need to keep it write down.